### ISSUE TITLE
Sign Up/In/Out - Issues 8.24

### ISSUE THREAD

**Issue by limichelle21** [limichelle21/ecommerce#31]
**Owner-User (Devise)**
- [x] Sessions#Destroy -- the user session is destroyed, but does not auto redirect to root_path (Welcome#Index)

**Customer-User (custom Sessions controller)**
- [ ]~~Sessions#create action - undefined method `authenticate` on User. a Customer is retrieved correctly based on `find_by`. Fails on the `authenticate` method~~

```
def create
    user = User.find_by(email: params[:session][:email].downcase)

    if user && user.authenticate(params[:session][:password])
      create_session(user)
      flash[:notice] = "Welcome, #{user.name}!"
      redirect_to store_path(store_id: current_store.id)
    else
      flash.now[:alert] = 'Invalid email/password combination'
      render :new
    end
  end

```

**Comment by limichelle21** [comment]
`destroy_user_session_path(current_user)` was appending the current_user to the link. Removed `(current_user)` and sign out redirect worked

**Comment by limichelle21** [comment]
Customer-signup => create a new Customer, save, and create new session for @customer

Output for original attempt that created a session/updated navbar with appropriate info

```
Started POST "/21/users/sign_up" for ::1 at 2016-08-25 16:30:57 +0800
Processing by CustomersController#create as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"40H2/HuwtIehW2bfux/PaPAGd9SBAM1Da1zoqED5jna7b/NPu0ZOnwbBs0udonAvJ7yVr1rHUgIdDpyrUyu4Sg==", "customer"=>{"email"=>"limi@gmail.com", "password"=>"[FILTERED]", "password_confirmation"=>"[FILTERED]", "type"=>"Customer"}, "commit"=>"Sign up", "store_id"=>"21"}
  Store Load (0.3ms)  SELECT  "stores".* FROM "stores" WHERE "stores"."slug" = ?  ORDER BY "stores"."id" ASC LIMIT 1  [["slug", "21"]]
  Store Load (0.5ms)  SELECT  "stores".* FROM "stores" WHERE "stores"."id" = ? LIMIT 1  [["id", 21]]
  CACHE (0.0ms)  SELECT  "stores".* FROM "stores" WHERE "stores"."slug" = ?  ORDER BY "stores"."id" ASC LIMIT 1  [["slug", "21"]]
  CACHE (0.0ms)  SELECT  "stores".* FROM "stores" WHERE "stores"."id" = ? LIMIT 1  [["id", "21"]]
   (0.2ms)  begin transaction
  User Exists (0.1ms)  SELECT  1 AS one FROM "users" WHERE "users"."email" = 'limi@gmail.com' LIMIT 1
  SQL (0.7ms)  INSERT INTO "users" ("type", "email", "encrypted_password", "store_id", "created_at", "updated_at") VALUES (?, ?, ?, ?, ?, ?)  [["type", "Customer"], ["email", "limi@gmail.com"], ["encrypted_password", "$2a$10$g724mp1rX7TKp4l8DjQBou57PWQXBHe96yc.T2BU1cKHhihuj5CPu"], ["store_id", 21], ["created_at", "2016-08-25 08:30:58.067260"], ["updated_at", "2016-08-25 08:30:58.067260"]]
   (1.1ms)  commit transaction
   (0.1ms)  begin transaction
   (0.1ms)  commit transaction
  Store Load (0.1ms)  SELECT  "stores".* FROM "stores" WHERE "stores"."slug" = ?  ORDER BY "stores"."id" ASC LIMIT 1  [["slug", "21"]]
  Store Load (0.1ms)  SELECT  "stores".* FROM "stores" WHERE "stores"."id" = ? LIMIT 1  [["id", 21]]
Redirected to http://localhost:3000/21/products
Completed 302 Found in 375ms (ActiveRecord: 5.2ms)
```

All subsequent sign up attempts do not have the second set of 

```
  (0.1ms)  begin transaction
   (0.1ms)  commit transaction
```

**Comment by limichelle21** [comment]
`current_user` is a helper method available through default Devise
Need to set own current_user method for Customer-user sign up/in/out?

**Comment by limichelle21** [comment]
`authenticate` method does not work because `user` is running through default Devise, but the `authenticate` method only works with `bycrpt` gem and `has_secure_password` in Rails authentication

http://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html

But - seems like Devise uses `bycrpt` - should that mean Devise provides access to the bycrpt gem methods as well?
https://github.com/plataformatec/devise/blob/v3.5.2/devise.gemspec#L23

**Comment by limichelle21** [comment]
Opted to try and extend the Devise controllers for customers vs. custom controllers (customer#new, customer#create, and sessions#new, sessions#create, sessions#destroy)

`rails g devise:controllers customers`

routes.rb

```
 devise_for :customers, controllers: { 
      registrations: 'customers/registrations', 
      sessions: 'customers/sessions'
    }
```

new rake routes 

```
 Prefix Verb   URI Pattern                                             Controller#Action
                  new_user_session GET    /users/sign_in(.:format)                                users/sessions#new
                      user_session POST   /users/sign_in(.:format)                                users/sessions#create
              destroy_user_session DELETE /users/sign_out(.:format)                               users/sessions#destroy
                     user_password POST   /users/password(.:format)                               devise/passwords#create
                 new_user_password GET    /users/password/new(.:format)                           devise/passwords#new
                edit_user_password GET    /users/password/edit(.:format)                          devise/passwords#edit
                                   PATCH  /users/password(.:format)                               devise/passwords#update
                                   PUT    /users/password(.:format)                               devise/passwords#update
          cancel_user_registration GET    /users/cancel(.:format)                                 users/registrations#cancel
                 user_registration POST   /users(.:format)                                        users/registrations#create
             new_user_registration GET    /users/sign_up(.:format)                                users/registrations#new
            edit_user_registration GET    /users/edit(.:format)                                   users/registrations#edit
                                   PATCH  /users(.:format)                                        users/registrations#update
                                   PUT    /users(.:format)                                        users/registrations#update
                                   DELETE /users(.:format)                                        users/registrations#destroy
                    dashboard_root GET    /dashboard(.:format)                                    dashboard/products#index
          dashboard_store_products GET    /dashboard/stores/:store_id/products(.:format)          dashboard/products#index
                                   POST   /dashboard/stores/:store_id/products(.:format)          dashboard/products#create
       new_dashboard_store_product GET    /dashboard/stores/:store_id/products/new(.:format)      dashboard/products#new
      edit_dashboard_store_product GET    /dashboard/stores/:store_id/products/:id/edit(.:format) dashboard/products#edit
           dashboard_store_product GET    /dashboard/stores/:store_id/products/:id(.:format)      dashboard/products#show
                                   PATCH  /dashboard/stores/:store_id/products/:id(.:format)      dashboard/products#update
                                   PUT    /dashboard/stores/:store_id/products/:id(.:format)      dashboard/products#update
                                   DELETE /dashboard/stores/:store_id/products/:id(.:format)      dashboard/products#destroy
            dashboard_store_orders GET    /dashboard/stores/:store_id/orders(.:format)            dashboard/orders#index
             dashboard_store_order GET    /dashboard/stores/:store_id/orders/:id(.:format)        dashboard/orders#show
                                   DELETE /dashboard/stores/:store_id/orders/:id(.:format)        dashboard/orders#destroy
       dashboard_store_order_lines GET    /dashboard/stores/:store_id/order_lines(.:format)       dashboard/order_lines#index
        dashboard_store_order_line GET    /dashboard/stores/:store_id/order_lines/:id(.:format)   dashboard/order_lines#show
         dashboard_store_customers GET    /dashboard/stores/:store_id/customers(.:format)         dashboard/customers#index
                  dashboard_stores POST   /dashboard/stores(.:format)                             dashboard/stores#create
               new_dashboard_store GET    /dashboard/stores/new(.:format)                         dashboard/stores#new
                              root GET    /                                                       welcome#index
                             store GET    /:store_id(.:format)                                    stores#show
        new_customer_store_session GET    /customers/:store_id/sign_in(.:format)                  customers/sessions#new
            customer_store_session POST   /customers/:store_id/sign_in(.:format)                  customers/sessions#create
    destroy_customer_store_session DELETE /customers/:store_id/sign_out(.:format)                 customers/sessions#destroy
           customer_store_password POST   /customers/:store_id/password(.:format)                 devise/passwords#create
       new_customer_store_password GET    /customers/:store_id/password/new(.:format)             devise/passwords#new
      edit_customer_store_password GET    /customers/:store_id/password/edit(.:format)            devise/passwords#edit
                                   PATCH  /customers/:store_id/password(.:format)                 devise/passwords#update
                                   PUT    /customers/:store_id/password(.:format)                 devise/passwords#update
cancel_customer_store_registration GET    /customers/:store_id/cancel(.:format)                   customers/registrations#cancel
       customer_store_registration POST   /customers/:store_id(.:format)                          customers/registrations#create
   new_customer_store_registration GET    /customers/:store_id/sign_up(.:format)                  customers/registrations#new
  edit_customer_store_registration GET    /customers/:store_id/edit(.:format)                     customers/registrations#edit
                                   PATCH  /customers/:store_id(.:format)                          customers/registrations#update
                                   PUT    /customers/:store_id(.:format)                          customers/registrations#update
                                   DELETE /customers/:store_id(.:format)                          customers/registrations#destroy
                        store_cart GET    /:store_id/cart(.:format)                               carts#show
                    store_products GET    /:store_id/products(.:format)                           products#index
                     store_product GET    /:store_id/products/:id(.:format)                       products#show
                      store_orders GET    /:store_id/orders(.:format)                             orders#index
                       store_order GET    /:store_id/orders/:id(.:format)                         orders#show
                store_confirmation GET    /:store_id/orders/:id/confirmation(.:format)            orders#confirmation
                 store_order_lines POST   /:store_id/order_lines(.:format)                        order_lines#create
                  store_order_line PATCH  /:store_id/order_lines/:id(.:format)                    order_lines#update
                                   PUT    /:store_id/order_lines/:id(.:format)                    order_lines#update
                                   DELETE /:store_id/order_lines/:id(.:format)                    order_lines#destroy
                     store_charges POST   /:store_id/charges(.:format)                            charges#create
                  new_store_charge GET    /:store_id/charges/new(.:format)                        charges#new
                                   GET    /:id(.:format)                                          stores#show
```

**Comment by mjthompsgb** [comment]
Hi Michelle,

Looks like you had some good ideas with Devise and bcrypt because Devise does use bcrypt.

As for extending Devise for customers it looks like 2 days ago you pasted those routes and looks like you may have got it working? I can see there is a CustomerController that appears to use very similar actions to UserController that is your first Devise controller. So is CustomerController extending the methods from Devise? I looked on Github and looks like your last push was 4 days ago so thinking this stuff isn't out there yet?

Let me know how it worked and if you have questions on it. I know sometimes extending Devise can be fun and I've never used it in a situation like this for different user types logging in from different places of the app.

**Comment by limichelle21** [comment]
Hi Matt,

Thanks for your notes! I think I got it working.. at least partially (a customer can Sign Up, Sign In, and Sign Out). 

I've also added a Membership model that connects customers and stores, so my registrations controller CREATE includes the following: 

customers/registrations_controller.rb 

```
def create
    super
    @customer = current_customer
    @customer.stores << current_store
    @customer.save
  end
```

I tested this a few days ago to add a `store` to `customer`, and in my rails c, it looks like it worked. 

As for routing, I have the following:

```
 scope '/:store_id' do
    devise_for :customers, controllers: {
      registrations: 'customers/registrations',
      sessions: 'customers/sessions'
    }
  end
```

Is that enough require a `store_id` when a customer logs in?

Thanks!
